### DISCUSSION TITLE
Creating static meshes programmatically

### DISCUSSION THREAD

**Discussion by schotime** [nxrighthere/UnrealCLR#56] [Help] [answered]
We need to generate static meshes programmatically, and would like to do it in c#.
There are possibly two ways to do it as far as I can see.

1. Invoke a CLR function that uses the managed API's to generate the mesh and add to the world
2. Return an array of points (via variables or other mechanism) to the blueprint and generate from within there.

Let me know if there are other ways or you have a better suggestion. 
Thx.

**Comment by nxrighthere** [comment]
I'll look into this after I finish relocation and IRL stuff.

**Reply by schotime** [reply]
No probs. Good luck with the move. 👍

**Reply by nxrighthere** [reply]
@schotime I'm still busy IRL but started to dig into this. So you would like to use [`UProceduralMeshComponent`](https://docs.unrealengine.com/en-US/API/Plugins/ProceduralMeshComponent/UProceduralMeshComponent/index.html) as far as I understand to generate meshes at runtime? From what I read, people prefer [this](https://github.com/TriAxis-Games/RuntimeMeshComponent) custom implementation since it has more features and probably works better in terms of performance. Have you tried it?

**Reply by jimboc1** [reply]
We haven't tried it, but we came to same conclusion - that RMC is the way to go, especially as Epic recently granted it and Epic loves people supporting granted projects and gives them grants.  😉 

Please let us know if there's anything we can do to help.

**Reply by nxrighthere** [reply]
Right. So after you give it a try, I'd like to know what you would like to have in the UnrealCLR API for your particular project to generate meshes from managed code, if required.

**Reply by jimboc1** [reply]
Shall do, thanks. I spent a bit of time today going through it - it's quite impressive. I'll do a few examples of what we initially wish to do via C++ and then let you know what we'd love to see in the UnrealCLR API.

**Comment by IceReaper** [comment]
Joining in. Any possible way to create map geometry and load models / materials / ... from filesystem via c# and using the UE for rendering only is highly appreciated.

**Reply by nxrighthere** [reply]
My goal would be to provide interoperability with [`RuntimeMeshComponent`](https://github.com/TriAxis-Games/RuntimeMeshComponent), because from what I see, this is what people prefer over built-in [`UProceduralMeshComponent`](https://docs.unrealengine.com/en-US/API/Plugins/ProceduralMeshComponent/UProceduralMeshComponent/index.html). I need some feedback about it to get an idea of how I can help.

**Comment by jimboc1** [comment, marked as answer]
An update on this - @schotime and I have now implemented dynamic mesh creation in C++ with the RMC sdk. It's working well and very promising. I could list some required bindings now, but as you indicated you've got a move in progress and will be time challenged until January, we thought we continue implementing to discover more of what we'd need rather than continually bugging you for more bindings. For example, we've done mesh creation, material assignment, UVs and manual LODs, but not colliders or physics.

But please let us know if you'd prefer to get more of a MVP implementation going earlier so you can indicate RMC compatibility, and we'll start implementing alongside in c#.

Thanks

**Reply by nxrighthere** [reply]
That's pretty cool! Can you please point me to API that you are using at the moment?

Also, is there any potential for interoperability through blueprints (once .NET blueprints evolve into a more robust and flexible implementation).

**Reply by IceReaper** [reply]
Would this allow to load and parse raw gltf models and custom map formats in c# and spawn in the meshes (with animations etc)? When speaking about materials: What about textures? can we push rgba byte[] from c# code to populate them?

**Reply by jimboc1** [reply]
The codebase is at https://github.com/TriAxis-Games/RuntimeMeshComponent .
No automatically generated API docs, but an overview at https://runtimemesh.koderz.io/usage.html .
I mainly went by the example project at https://github.com/TriAxis-Games/RuntimeMeshComponent-Examples

Yes, definitely blueprint interoperability potential, quite desirable. Currently we have defined in C++ a custom mesh provider (based on RMC's RuntimeMeshProvider) which sets up all the vertices/triangles/uvs/...  We have a blueprint based on a RuntimeMeshActor defined in UE4 which we drag into the scene. Its construction script creates an instance of our custom mesh provider, and we pass in things like materials to our mesh provider via properties/methods we expose to blueprints. Not that we have to use blueprints but it has made just made getting things going quicker.

**Reply by jimboc1** [reply]
> Would this allow to load and parse raw gltf models and custom map formats in c# and spawn in the meshes (with animations etc)? When speaking about materials: What about textures? can we push rgba byte[] from c# code to populate them?

Parsing model formats you'd either use what UE4 has built in or use another plugin on top such as Rama's https://github.com/subramanya1997/Import-3D-models-ingame , but the only animation RMC supports is by giving it new vertices/triangles to display. Any sort of glTF/Albembic/FBX/bone animation would be a higher layer on top of RMC. Same with textures, RMC just specialises in procedural mesh creation.

**Reply by nxrighthere** [reply]
Thanks for the info!

I think it's possible to make `RuntimeMeshComponent` work with 3rd-party libraries like [SharpGLTF](https://github.com/vpenades/SharpGLTF) and so on.

As for textures importing at runtime and assigning them to materials, it's supported with [`Texture2D`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/Texture2D.md) API. See constructors for creation from a raw PNG, JPEG, BMP, or EXR byte buffer or file.

**Reply by schotime** [reply]
hey @nxrighthere, just following this issue up. Is there a way to get some of these API's exposed so we could expose more or this is something you're likely to look at soon. We'd be happy to help out on the implementation if that were possibly once the base had been done, to give us something to work from, but this C++ thing is really slowing us down now, so it would be good to have a solution without it. thx.

**Reply by nxrighthere** [reply]
@schotime You mean exposing `UProceduralMeshComponent`? Please, provide more information on where you get stuck in your pipeline.

**Reply by schotime** [reply]
I mean RMC, but at the moment its these.

- RuntimeMeshComponent,
- Interactive Tools Framework
- MeshModelingToolset plugin

**Reply by nxrighthere** [reply]
Exposing functionality of 3rd-party plugins to managed code is problematic and I have certain concerns about maintainability and future updates of the engine which often breaks compatibility. This is a low-priority task for now.

What might help as a general-purpose solution is using blueprints as a glue between plugins and managed code, but https://github.com/nxrighthere/UnrealCLR/issues/33 should be finished for this.
